### PR TITLE
[DCP] overwrites existing checkpoint by default

### PR DESCRIPTION
**Pull request by LucasLLC**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125877

Checks for existing checkpoints and overwrites, based on an `overwrite` flag

Differential Revision: [D57186174](https://our.internmc.facebook.com/intern/diff/D57186174/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: test/distributed/checkpoint/test_fsspec.py [hunk 1]
```diff
@@ -176,6 +177,25 @@ def opt_at(opt, idx):
...
+        dcp.save({"random": t1}, checkpoint_id=self.temp_dir)
+        dcp.save({"random": t2}, checkpoint_id=self.temp_dir)
+
+        sd = {"random": torch.zeros(10)}
+        dcp.load(sd, checkpoint_id=self.temp_dir)
+
+        self.assertTrue(torch.allclose(sd["random"], t2))
+
+        dcp.save(
+            {"random": t2}, storage_writer=FsspecWriter(self.temp_dir, overwrite=True)
```
Comment (wz337) [inline test/distributed/checkpoint/test_fsspec.py:195]: Should this be overwrite=False when exception is raised, As the default is overwrite=True.

Comment (LucasLLC) [reply test/distributed/checkpoint/test_fsspec.py:195]: yes, and the format for assert raises is wrong as well, I'll have to take a look

Comment (LucasLLC) [reply test/distributed/checkpoint/test_fsspec.py]: fixed, thanks @wz337 !

**Review by fegin — COMMENTED** [review]
It's good to have the overwritten feature. But shouldn't this be an opt-in feature? Just like the `cp` linux command will explicitly ask if users would want to overwrite the existing checkpointing. If we do this implicitly, would it cause some surprises to the users?

**Comment by LucasLLC** [comment]
> It's good to have the overwritten feature. But shouldn't this be an opt-in feature? Just like the `cp` linux command will explicitly ask if users would want to overwrite the existing checkpointing. If we do this implicitly, would it cause some surprises to the users?

good point, will update

**Comment by LucasLLC** [comment]
Will fix errors before merging

**Comment by LucasLLC** [comment]
> Will fix errors before merging



> > It's good to have the overwritten feature. But shouldn't this be an opt-in feature? Just like the `cp` linux command will explicitly ask if users would want to overwrite the existing checkpointing. If we do this implicitly, would it cause some surprises to the users?
> 
> good point, will update

We decided actually that setting `overwrite=True` as the default is safer due to BC considerations, since it turns out depending on the implementation (fsspec vs. filesystem), overwrite may or may not error when attempting to overwrite, but in the case of this PR will always raise if `overwrite=False`

I added a warning when we detect a checkpoint, and will change `overwrite=False` as the default after 2.5

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**